Migrate to docusaurus (#668)
iotaledger/wallet.rs: docOps/docs/libraries/nodejs/examples.md

The following example creates a new database and account:
```javascript
require('dotenv').config()

async function run() {
    const { AccountManager, SignerType } = require('@iota/wallet')
    const manager = new AccountManager({
        storagePath: './alice-database',
    })
    manager.setStrongholdPassword(process.env.SH_PASSWORD)
    manager.storeMnemonic(SignerType.Stronghold)

    const account = await manager.createAccount({
        clientOptions: { node: "https://api.lb-0.testnet.chrysalis2.com", localPow: true },
        alias: 'Alice',
    })

    console.log('Account created:', account.alias())
      
}

run()
```
* Storage is initialized under the given path (`./alice-database`)
* The password is set based on your password in _.env_ file ( _manager.setStrongholdPassword(process.env.SH_PASSWORD)_ )
* When you initialize the new database, a Stronghold mnemonic (seed) is automatically generated and stored by default ( _manager.storeMnemonic(SignerType.Stronghold)_ ).
* The seed should be set only for the first time. In order to open already initialized database, you can simply use your password.

> The storage is encrypted at rest, so you need a strong password and location where to place your storage. 

Reviewer: section-wide: maybe "encrypted while not in use" for extra clarity.

Author: I think encrypted at rest is quite clear, and how this sort of mechanism is usually called.

:::

 The storage comprises two things:
* A single file called _wallet.stronghold_ , which contains _seed_.  `Stronghold` will secure the seed and encrypt it at rest. The generated seed (mnemonic) serves as a cryptographic key, which is used to generate all accounts and related addresses.
* Other data used by library that is stored under _db_ sub-directory.  The includes account information, generated addresses, fetched messages, etc. This data is used to speed up some operations, such as account creation, address generation, etc.

One of the key principles behind `Stronghold` based storage is that no one can extract a seed from the storage. You deal with all accounts purely via an _AccountManager_ instance and all complexities are hidden under the hood and are dealt with securely.

If you also want to store a seed somewhere else, you can use the _AccountManager.generateMnemonic()_ method. You can use this method to generate a random seed.  You can also use it before the actual account initialization.

You can find detailed information about seed generation at [Developer Guide to Chrysalis](https://chrysalis.docs.iota.org/guides/dev_guide.html#seed).

> ### Accounts

Reviewer: ```suggestion
### Accounts

```

> The `Wallet.rs` library uses a model of individual accounts to separate individual users/clients from each other. It is possible to generate multiple addresses for each account deterministically. 

Reviewer: Why would I choose one over the other?

```javascript
 require('dotenv').config();

async function run() {
	const { AccountManager, RemainderValueStrategy } = require('@iota/wallet')
    const manager = new AccountManager({
        storagePath: './alice-database'
    })

    manager.setStrongholdPassword(process.env.SH_PASSWORD)

    const account = manager.getAccount('Alice')
    
    console.log('alias', account.alias())
    console.log('syncing...')
    const synced = await account.sync()
    console.log('available balance', account.balance().available)
    
    //TODO: Replace with the address of your choice!
	const addr = 'atoi1qykf7rrdjzhgynfkw6z7360avhaaywf5a4vtyvvk6a06gcv5y7sksu7n5cs'
	const amount = 10000000

	const node_response = await account.send(
		addr,
		amount,
        {remainderValueStrategy: RemainderValueStrategy.reuseAddress()}
    ) 

    console.log(`Check your message on https://explorer.iota.org/chrysalis/message/${node_response.id}`)
}

run()
```
The full function signature is `Account.send(address, amount, [options])`.
You can use the default options. However, you can provide additional options, such as _remainderValueStrategy_ which has the following strategies:

> * _changeAddress()_ : Send the remainder value to an internal address

Reviewer: For function names, change *emphasis* to the `inline codeblock`

* _reuseAddress()_ : Send the remainder value back to its original address

The _Account.send()_ function returns a _wallet message_ that fully describes the given transaction. You can use the _messageId_ to check confirmation status. You can retrieve individual messages related to any given account using the _Account.listMessages()_ function.

> ### Dust protection

Reviewer: ```suggestion
### Dust Protection

```

The network uses a [dust protection](https://chrysalis.docs.iota.org/guides/dev_guide.html#dust-protection) protocol to prevent malicious actors from spamming the network while also keeping track of the unspent amount ( _UTXO_ ).

:::info
Micro-transaction below 1Mi of IOTA tokens can be sent to another address if there is already at least 1Mi on that address. 
That's why we sent 1Mi in the last example to comply with the dust protection.
:::

> Dust protection also means you can't leave less than 1Mi on a spent address (leave a dust behind).

Reviewer: I thought that only outputs can be spent, not addresses, do I miss something?

also, it belongs to the infobox.

## Backup a database

> Due to security practices that are incorporated in the `Stronghold's` DNA, there's no way to retrieve a seed, as it is encrypted at rest.  Therefore, if you're using the default options,  backing up the seed storage is a very important task. 

Reviewer: why inline codeblock?

No contractions please.

Replace the last sentence with a call to action (so make a backup of your seed).

The following example will guide you in backing up your data in secure files. You can move this file to another app or device, and restore it.

```javascript
require('dotenv').config();

async function run() {

    const { AccountManager } = require('@iota/wallet')
    const manager = new AccountManager({
        storagePath: './alice-database'
    })

    manager.setStrongholdPassword(process.env.SH_PASSWORD)

    let backup_path = await manager.backup("./backup", process.env.SH_PASSWORD)
    
    console.log('Backup path:', backup_path)
}

run()
```
Alternatively, you can create a copy of the _wallet.stronghold_ file and use it as seed backup. This can be achieved by a daily [_cronjob_](https://linux.die.net/man/1/crontab), [_rsync_](https://linux.die.net/man/1/rsync) or [_scp_](https://linux.die.net/man/1/scp) with a datetime suffix for example.

> ## Restore database

Reviewer: ```suggestion
## Restore database

```

To restore a database via `Wallet.rs`, you will need to:
1. Create new empty database with a password (without mnemonic seed)

> 2. Import all accounts from the file that has been backed up earlier

Reviewer: You can explain it in a plain paragraph.

The following example restores a secured backup file:
```javascript
require('dotenv').config();

async function run() {

    const { AccountManager } = require('@iota/wallet')
    const manager = new AccountManager({
        storagePath: './alice-database'
    })

    // Add the path to the file from example 5-backup.js
    // for example: ./backup/2021-02-12T01-23-11-iota-wallet-backup-wallet.stronghold
    let backup_path = "input your backup file"

    await manager.importAccounts(backup_path, process.env.SH_PASSWORD)
    const account = manager.getAccount('Alice')
    console.log('Account:', account.alias())
}

run()
```

Since the backup file is just a copy of the original database it can be also be renamed to _wallet.stronghold_ and opened in a standard way.

> ## Listening to events

Reviewer: ```suggestion
## Listening to events

```

    console.log('Account:', account.alias())

    // Always sync before doing anything with the account
    const synced = await account.sync()
    console.log('Syncing...')
    // let address = account.generateAddress()

    // get latest address
    let addressObject = account.latestAddress()

    console.log("Address:", addressObject.address)

    // Use the Chrysalis Faucet to send testnet tokens to your address:
    console.log("Fill your address with the Faucet: https://faucet.testnet.chrysalis2.com/")

    const callback = function (err, data) {
        console.log("data:", data)
    }

    addEventListener("BalanceChange", callback)

    // Possible Event Types:
    //
    // ErrorThrown
    // BalanceChange
    // NewTransaction
    // ConfirmationStateChange
    // Reattachment
    // Broadcast
}

run()
```

Example output:

```json
data: {
  accountId: 'wallet-account://1666fc60fc95534090728a345cc5a861301428f68a237bea2b5ba0c844988566',
  address: {
    address: 'atoi1q9c6r2ek5w2yz54en78m8dxwl4qmwd7gmh9u0krm45p8txxyhtfry6apvwj',
    balance: 20000000,
    keyIndex: 0,
    internal: false,
    outputs: [ [Object], [Object] ]
  },
  balance: 20000000
}
```

> You can then use the _accountId_ to identify the account via _AccountManager.getAccount(accountId)_.

Reviewer: ```suggestion
You can then use the _accountId_ to identify the account via `AccountManager.getAccount(accountId)`.
```

Read more about Events in the [API reference](api_reference.md#addeventlistenerevent-cb).

> 

Reviewer: ```suggestion
```

> ## Migration 

Reviewer: ```suggestion
## Migration 

```

# Examples

This section will guide you through several examples using the node.js binding of the `Wallet.rs` library. You can also find the code for the examples in the `/bindings/nodejs/examples` folder in the [official GitHub repository](https://github.com/iotaledger/wallet.rs/tree/develop/bindings/nodejs/examples).

> All the examples in this section expect your custom password  to be set in the  _.env_  file:

Reviewer: passive voice to active voice

```bash
SH_PASSWORD="here is your super secure password"
```

## Security
:::warning
It is not recommended to store passwords on the host's environment variables, or in the source code in a production setup. 

> Please make sure you follow our [backup and security recommendations](https://chrysalis.docs.iota.org/guides/backup_security.html) for production use.

Reviewer: ```suggestion
In a production setup, do not store passwords in the host's environment variables or in the source code. See our [backup and security recommendations](https://chrysalis.docs.iota.org/guides/backup_security.html) for production setups.
```

```javascript
require('dotenv').config()

async function run() {
    const { AccountManager, SignerType } = require('@iota/wallet')
    const manager = new AccountManager({
        storagePath: './alice-database',
    })
    manager.setStrongholdPassword(process.env.SH_PASSWORD)
    manager.storeMnemonic(SignerType.Stronghold)

    const account = await manager.createAccount({
        clientOptions: { node: "https://api.lb-0.testnet.chrysalis2.com", localPow: true },
        alias: 'Alice',
    })

    console.log('Account created:', account.alias())
      
}

run()
```
* Storage is initialized under the given path (`./alice-database`)
* The password is set based on your password in  _.env_  file ( _manager.setStrongholdPassword(process.env.SH_PASSWORD)_ )
* When you initialize the new database, a Stronghold mnemonic (seed) is automatically generated and stored by default ( _manager.storeMnemonic(SignerType.Stronghold)_ ).
* The seed should be set only for the first time. In order to open already initialized database, you can simply use your password.

The storage is encrypted at rest, so you need a strong password and location where to place your storage. 

:::warning

> We highly recommended you store your `Stronghold` password encrypted on rest and separated from `Stronghold` snapshots. 

Reviewer: to store

> Deal with the password with utmost care

Reviewer: dot

:::

> Technically speaking, the storage comprises two things:

Reviewer: Technically speaking is a buzzword.

* A single file called  _wallet.stronghold_ , which contains  _seed_  and is secured by `Stronghold` and encrypted at rest. The generated seed (mnemonic) serves as a cryptographic key from which all accounts and related addresses are generated.
* Other data used by library that is stored under  _db_  sub-directory.  The includes account information, generated addresses, fetched messages, etc. This data is used to speed up some operations, such as account creation, address generation, etc.

One of the key principles behind `Stronghold` based storage is that no one can extract a seed from the storage. You deal with all accounts purely via an  _AccountManager_  instance and all complexities are hidden under the hood and are dealt with securely.

> If you also want to store a seed somewhere else, you can use the  _AccountManager.generateMnemonic()_  method. This method will generate a random seed, and it can be used before the actual account initialization.

Reviewer: Lots of passive voice that is justified only in a few occurences.

> You can find detailed information about seed generation at [Developer guide to Chrysalis](https://chrysalis.docs.iota.org/guides/dev_guide.html#seed).

Reviewer: ```suggestion
You can find detailed information about seed generation at [developer guide to Chrysalis](https://chrysalis.docs.iota.org/guides/dev_guide.html#seed).
```

### Accounts
The `Wallet.rs` library uses a model of individual accounts to separate individual users/clients from each other. It is possible to generate multiple addresses for each account deterministically. 

Once the backend storage has been created, individual accounts for individual users can be created by running the  _manager.createAccount()_  method:

```javascript
    let account = await manager.createAccount({
        alias: 'Alice',  // an unique id from your existing user
        clientOptions: { node: 'http://api.lb-0.testnet.chrysalis2.com', localPow: false }
    })
```

> Each account is related to a specific IOTA network (mainnet / devnet), which is referenced by a node properties such as node url.  In this example, the `Chrysalis` testnet balancer.

Reviewer: is it devnet or testnet?

For more information about  _clientOptions_ , please refer to [Wallet NodeJs API Reference](api_reference.md).

 _Alias_  should be unique, and it can be any string that you see fit. The  _alias_  is usually used to identify the account later on. Each account is also represented by an  _index_  which is incremented by 1 every time new account is created. 
Any account can be then referred to by its  _index_ ,  _alias_  or one of its generated  _addresses_ .

Several API calls can be performed via an  _account_  instance.

:::info

> It is a good practice to sync the given account with the Tangle every time you work with an  _account_  instance to rely on the latest information available.  You can do this using _account.sync()_. By default, _account.sync()_ is performed automatically on _send_, _retry_, _reattach_ and _promote_ API calls.:::

Reviewer: Why a good practice?

*  _account.listAddresses()_ : returns list of addresses related to the account.
*  _account.getUnusedAddress()_ : returns a first unused address.
*  _account.generateAddress()_ : generate a new address for the address index incremented by 1.
*  _account.balance()_ : returns the balance for the given account.
*  _account.sync()_ : sync the account information with the tangle.

## Generating Address(es)
Each account can have multiple addresses. Addresses are generated deterministically based on the account and address index. This means that the combination of account and index uniquely identifies the given address.

There are two types of addresses, _internal_ and _public_ (external), and each set of addresses is independent of each other and has independent  _index_  id.

* _Public_ addresses are created by _account.generateAddress()_ and  are indicated as _internal=false_ (public)
* _Internal_ addresses are also called  _change_  addresses. _Internal_ addresses are used to store the excess funds and are indicated as _internal=false_.

This approach is also known as a *BIP32 Hierarchical Deterministic wallet (HD Wallet)*.

:::info

>  You may remember IOTA 1.0 network in which addresses were not reusable. This is no longer true and addresses can be reused multiple times in the IOTA 1.5 (Chrysalis) network._

Reviewer: Probably just say that you can reuse addresses in Crysalis in a plain paragraph without referring to good old times.

    // console.log("Address:", addressObject.address)

    // Use the Chrysalis Faucet to send testnet tokens to your address:
    console.log("Fill your address with the Faucet: https://faucet.testnet.chrysalis2.com/")
}

run()
```

## Checking Balance
Before we continue further, please visit the [IOTA testnet faucet service](https://faucet.testnet.chrysalis2.com/) and send to your testnet addresses some tokens.

![IOTA Faucet Service](../../../static/img/libraries/screenshot_faucet.png)

You can use the following example to generate a new database and account:

```javascript

require('dotenv').config()

async function run() {
	const { AccountManager } = require('@iota/wallet')
    const manager = new AccountManager({
        storagePath: './alice-database'
    })

    manager.setStrongholdPassword(process.env.SH_PASSWORD)

    const account = manager.getAccount('Alice')
    
    console.log('Account:', account.alias())
    
    // Always sync before doing anything with the account
    const synced = await account.sync()
    console.log('Syncing...')

    console.log('Available balance', account.balance().available)
}

run()
```

> IOTA is based on _Unspent Transaction Output_ model. You can find a detailed explanation in the [Developer guide to Chrysalis](https://chrysalis.docs.iota.org/guides/dev_guide.html#unspent-transaction-output-utxo).

Reviewer: How exactly "Developer guide to Chrysalis" should be capitalized? :confused:

## Sending tokens

> You can use the following example to send tokens using an  _Account_  instance to any desired  _address_ :

Reviewer: I don't think we need emphasis here.

```javascript
 require('dotenv').config();

async function run() {
	const { AccountManager, RemainderValueStrategy } = require('@iota/wallet')
    const manager = new AccountManager({
        storagePath: './alice-database'
    })

    manager.setStrongholdPassword(process.env.SH_PASSWORD)

    const account = manager.getAccount('Alice')
    
    console.log('alias', account.alias())
    console.log('syncing...')
    const synced = await account.sync()
    console.log('available balance', account.balance().available)
    
    //TODO: Replace with the address of your choice!
	const addr = 'atoi1qykf7rrdjzhgynfkw6z7360avhaaywf5a4vtyvvk6a06gcv5y7sksu7n5cs'
	const amount = 10000000

	const node_response = await account.send(
		addr,
		amount,
        {remainderValueStrategy: RemainderValueStrategy.reuseAddress()}
    ) 

    console.log(`Check your message on https://explorer.iota.org/chrysalis/message/${node_response.id}`)
}

run()
```
The full function signature is `Account.send(address, amount, [options])`.

> You can use the default options, however additional options can be provided, such as  _remainderValueStrategy_  which has the following options:

Reviewer: Split the sentence.

*  _changeAddress()_ : Send the remainder value to an internal address
*  _reuseAddress()_ : Send the remainder value back to its original address

The  _Account.send()_  function returns a  _wallet message_  that fully describes the given transaction. You can use the  _messageId_  to check confirmation status. You can retrieve individual messages related to any given account using the  _Account.listMessages()_  function.

### Dust protection
The network uses a [dust protection](https://chrysalis.docs.iota.org/guides/dev_guide.html#dust-protection) protocol to prevent malicious actors from spamming the network while also keeping track of the unspent amount ( _UTXO_ ).

:::info

> "... micro-transaction below 1Mi of IOTA tokens can be sent to another address if there is already at least 1Mi on that address. 

Reviewer: If that's a quote, why infobox and not a quote? What's the source?

That's why we sent 1Mi in the last example to comply with the protection."
:::

Dust protection also means you can't leave less than 1Mi on a spent address (leave a dust behind).

## Backup a database

Due to security practices that are incorporated in the `Stronghold's` DNA, there's no way to retrieve a seed, as it is encrypted at rest.  Therefore, if you're using the default options,  backing up the seed storage is a very important task. 

The following example will guide you in backing up your data in secure files. You can move this file to another app or device, and restore it.

```javascript
require('dotenv').config();

async function run() {

    const { AccountManager } = require('@iota/wallet')
    const manager = new AccountManager({
        storagePath: './alice-database'
    })

    manager.setStrongholdPassword(process.env.SH_PASSWORD)

    let backup_path = await manager.backup("./backup", process.env.SH_PASSWORD)
    
    console.log('Backup path:', backup_path)
}

run()
```

> Alternatively, you can create a copy of the  _wallet.stronghold_  file and use it as seed backup. This can be achieved by a daily _cronjob_, _rsync_ or _scp_ with a datetime suffix for example.

Reviewer: Link docu for all of that!

    // get latest address
    let addressObject = account.latestAddress()

    console.log("Address:", addressObject.address)

    // Use the Chrysalis Faucet to send testnet tokens to your address:
    console.log("Fill your address with the Faucet: https://faucet.testnet.chrysalis2.com/")

    const callback = function (err, data) {
        console.log("data:", data)
    }

    addEventListener("BalanceChange", callback)

    // Possible Event Types:
    //
    // ErrorThrown
    // BalanceChange
    // NewTransaction
    // ConfirmationStateChange
    // Reattachment
    // Broadcast
}

run()
```

Example output:

```json
data: {
  accountId: 'wallet-account://1666fc60fc95534090728a345cc5a861301428f68a237bea2b5ba0c844988566',
  address: {
    address: 'atoi1q9c6r2ek5w2yz54en78m8dxwl4qmwd7gmh9u0krm45p8txxyhtfry6apvwj',
    balance: 20000000,
    keyIndex: 0,
    internal: false,
    outputs: [ [Object], [Object] ]
  },
  balance: 20000000
}
```

You can then use the _accountId_ to identify the account via _AccountManager.getAccount(accountId)_.

Read more about Events in the [API reference](api_reference.md#addeventlistenerevent-cb).

## Migration 

> You can use the following example to create a new database and account, and migrate funds from the legacy network to the `Chrysalis` network.

Reviewer: This example is so big so it deserves a separate article with step-by-step explanation.

Author: I agree on this.  Perhaps even a whole migration section in which we address this for all the bindings, though I'm not sure it falls within the scope of this task.
